File: constants/system.ts
// Critical system constants extracted to break circular dependencies

import { feature } from 'bun:bundle'
import { getFeatureValue_CACHED_MAY_BE_STALE } from '../services/analytics/growthbook.js'
import { isCustomBackendEnabled } from '../utils/customBackend.js'
import { logForDebugging } from '../utils/debug.js'
import { isEnvDefinedFalsy } from '../utils/envUtils.js'
import { getAPIProvider } from '../utils/model/providers.js'
import { getWorkload } from '../utils/workloadContext.js'
import { getProductAssistantName, getProductDisplayName } from './product.js'

const DEFAULT_PREFIX = `You are Mossen, a software engineering assistant running in the Mossen CLI.`
const MOSSEN_AGENT_SDK_PRESET_PREFIX = `You are Mossen, a software engineering assistant running within the Mossen Agent SDK.`
const MOSSEN_AGENT_SDK_PREFIX = `You are a Mossen agent, built on the Mossen Agent SDK.`
const CUSTOM_DEFAULT_PREFIX = `You are ${getProductAssistantName()}, a software engineering assistant running in a local ${getProductDisplayName()} environment.`
const CUSTOM_AGENT_SDK_PRESET_PREFIX = `You are ${getProductAssistantName()} running inside a local ${getProductDisplayName()} agent runtime.`
const CUSTOM_AGENT_SDK_PREFIX = `You are a software engineering agent operating through ${getProductDisplayName()}'s custom model backend.`

const CLI_SYSPROMPT_PREFIX_VALUES = [
  DEFAULT_PREFIX,
  MOSSEN_AGENT_SDK_PRESET_PREFIX,
  MOSSEN_AGENT_SDK_PREFIX,
  CUSTOM_DEFAULT_PREFIX,
  CUSTOM_AGENT_SDK_PRESET_PREFIX,
  CUSTOM_AGENT_SDK_PREFIX,
] as const

export type CLISyspromptPrefix = (typeof CLI_SYSPROMPT_PREFIX_VALUES)[number]

/**
 * All possible CLI sysprompt prefix values, used by splitSysPromptPrefix
 * to identify prefix blocks by content rather than position.
 */
export const CLI_SYSPROMPT_PREFIXES: ReadonlySet<string> = new Set(
  CLI_SYSPROMPT_PREFIX_VALUES,
)

export function getCLISyspromptPrefix(options?: {
  isNonInteractive: boolean
  hasAppendSystemPrompt: boolean
}): CLISyspromptPrefix {
  if (isCustomBackendEnabled()) {
    if (options?.isNonInteractive) {
      if (options.hasAppendSystemPrompt) {
        return CUSTOM_AGENT_SDK_PRESET_PREFIX
      }
      return CUSTOM_AGENT_SDK_PREFIX
    }
    return CUSTOM_DEFAULT_PREFIX
  }
  const apiProvider = getAPIProvider()
  if (apiProvider === 'vertex') {
    return DEFAULT_PREFIX
  }

  if (options?.isNonInteractive) {
    if (options.hasAppendSystemPrompt) {
        return MOSSEN_AGENT_SDK_PRESET_PREFIX
      }
      return MOSSEN_AGENT_SDK_PREFIX
    }
    return DEFAULT_PREFIX
  }

/**
 * Check if attribution header is enabled.
 * Enabled by default, can be disabled via env var or GrowthBook killswitch.
 */
function isAttributionHeaderEnabled(): boolean {
  if (isEnvDefinedFalsy(process.env.MOSSEN_CODE_ATTRIBUTION_HEADER)) {
    return false
  }
  return getFeatureValue_CACHED_MAY_BE_STALE('tengu_attribution_header', true)
}

/**
 * Get attribution header for API requests.
 * Returns a header string with cc_version (including fingerprint) and cc_entrypoint.
 * Enabled by default, can be disabled via env var or GrowthBook killswitch.
 *
 * When NATIVE_CLIENT_ATTESTATION is enabled, includes a `cch=00000` placeholder.
 * Before the request is sent, Bun's native HTTP stack finds this placeholder
 * in the request body and overwrites the zeros with a computed hash. The
 * server verifies this token to confirm the request came from a real Mossen
 * Code client. See the native HTTP attestation implementation for details.
 *
 * We use a placeholder (instead of injecting from Zig) because same-length
 * replacement avoids Content-Length changes and buffer reallocation.
 */
export function getAttributionHeader(fingerprint: string): string {
  if (isCustomBackendEnabled()) {
    return ''
  }
  if (!isAttributionHeaderEnabled()) {
    return ''
  }

  const version = `${MACRO.VERSION}.${fingerprint}`
  const entrypoint = process.env.MOSSEN_CODE_ENTRYPOINT ?? 'unknown'

  // cch=00000 placeholder is overwritten by Bun's HTTP stack with attestation token
  const cch = feature('NATIVE_CLIENT_ATTESTATION') ? ' cch=00000;' : ''
  // cc_workload: turn-scoped hint so the API can route e.g. cron-initiated
  // requests to a lower QoS pool. Absent = interactive default. Safe re:
  // fingerprint (computed from msg chars + version only, line 78 above) and
  // cch attestation (placeholder overwritten in serialized body bytes after
  // this string is built). Server _parse_cc_header tolerates unknown extra
  // fields so old API deploys silently ignore this.
  const workload = getWorkload()
  const workloadPair = workload ? ` cc_workload=${workload};` : ''
  const header = `x-mossen-billing-header: cc_version=${version}; cc_entrypoint=${entrypoint};${cch}${workloadPair}`

  logForDebugging(`attribution header ${header}`)
  return header
}

File: crates/mossen-types/src/constants/system.rs
//! # System (system.ts)
//!
//! 关键系统常量，用于打破循环依赖。

use std::collections::HashSet;

use once_cell::sync::Lazy;

pub const DEFAULT_PREFIX: &str =
    "You are Mossen, a software engineering assistant running in the Mossen CLI.";
pub const MOSSEN_AGENT_SDK_PRESET_PREFIX: &str =
    "You are Mossen, a software engineering assistant running within the Mossen Agent SDK.";
pub const MOSSEN_AGENT_SDK_PREFIX: &str = "You are a Mossen agent, built on the Mossen Agent SDK.";

/// Custom backend default prefix template.
/// In TS: ``You are ${getProductAssistantName()}, a software engineering assistant running in a local ${getProductDisplayName()} environment.``
pub fn custom_default_prefix(assistant_name: &str, product_name: &str) -> String {
    format!(
        "You are {}, a software engineering assistant running in a local {} environment.",
        assistant_name, product_name
    )
}

/// Custom backend agent SDK preset prefix template.
pub fn custom_agent_sdk_preset_prefix(assistant_name: &str, product_name: &str) -> String {
    format!(
        "You are {} running inside a local {} agent runtime.",
        assistant_name, product_name
    )
}

/// Custom backend agent SDK prefix template.
pub fn custom_agent_sdk_prefix(product_name: &str) -> String {
    format!(
        "You are a software engineering agent operating through {}'s custom model backend.",
        product_name
    )
}

/// All possible CLI sysprompt prefix values (static ones).
/// Used by splitSysPromptPrefix to identify prefix blocks by content.
pub static CLI_SYSPROMPT_PREFIXES: Lazy<HashSet<&'static str>> = Lazy::new(|| {
    let mut s = HashSet::new();
    s.insert(DEFAULT_PREFIX);
    s.insert(MOSSEN_AGENT_SDK_PRESET_PREFIX);
    s.insert(MOSSEN_AGENT_SDK_PREFIX);
    // Note: custom backend prefixes are dynamic and checked separately at runtime
    s
});

/// Get the CLI sysprompt prefix based on context.
/// `is_custom_backend`: whether custom backend is enabled
/// `is_non_interactive`: whether the session is non-interactive
/// `has_append_system_prompt`: whether there's an appended system prompt
/// `api_provider`: the API provider string (e.g., "vertex")
/// `assistant_name`: product assistant name for custom backend
/// `product_name`: product display name for custom backend
pub fn get_cli_sysprompt_prefix(
    is_custom_backend: bool,
    is_non_interactive: bool,
    has_append_system_prompt: bool,
    api_provider: &str,
    assistant_name: &str,
    product_name: &str,
) -> String {
    if is_custom_backend {
        if is_non_interactive {
            if has_append_system_prompt {
                return custom_agent_sdk_preset_prefix(assistant_name, product_name);
            }
            return custom_agent_sdk_prefix(product_name);
        }
        return custom_default_prefix(assistant_name, product_name);
    }

    if api_provider == "vertex" {
        return DEFAULT_PREFIX.to_string();
    }

    if is_non_interactive {
        if has_append_system_prompt {
            return MOSSEN_AGENT_SDK_PRESET_PREFIX.to_string();
        }
        return MOSSEN_AGENT_SDK_PREFIX.to_string();
    }
    DEFAULT_PREFIX.to_string()
}

/// Check if attribution header is enabled.
/// `env_falsy`: whether MOSSEN_CODE_ATTRIBUTION_HEADER env var is defined and falsy.
/// `growthbook_value`: the GrowthBook `mossen_attribution_header` value (default true).
pub fn is_attribution_header_enabled(env_falsy: bool, growthbook_value: bool) -> bool {
    if env_falsy {
        return false;
    }
    growthbook_value
}

/// Get attribution header for API requests.
/// Returns a header string with cc_version (including fingerprint) and cc_entrypoint.
///
/// When native client attestation is enabled, includes a `cch=00000` placeholder.
/// Before the request is sent, Bun's native HTTP stack finds this placeholder
/// in the request body and overwrites the zeros with a computed hash.
pub fn get_attribution_header(
    is_custom_backend: bool,
    is_enabled: bool,
    version: &str,
    fingerprint: &str,
    entrypoint: &str,
    native_attestation: bool,
    workload: Option<&str>,
) -> String {
    if is_custom_backend {
        return String::new();
    }
    if !is_enabled {
        return String::new();
    }

    let ver = format!("{}.{}", version, fingerprint);
    let cch = if native_attestation {
        " cch=00000;"
    } else {
        ""
    };
    let workload_pair = match workload {
        Some(w) => format!(" cc_workload={};", w),
        None => String::new(),
    };
    format!(
        "x-mossen-billing-header: cc_version={}; cc_entrypoint={};{}{}",
        ver, entrypoint, cch, workload_pair
    )
}
